examples/ex1.cpp

```diff
@@ -194,9 +194,12 @@ int main(int argc, char *argv[])
       umf_solver.Mult(B, X);
 #endif
    }
-   else // No preconditioning for now in partial assembly mode.
+   else // Jacobi preconditioning in partial assembly mode
    {
-      CG(*A, B, X, 1, 2000, 1e-12, 0.0);
+      Vector diag_pa(fespace->GetTrueVSize());
+      a->AssembleDiagonal(diag_pa);
+      OperatorJacobiSmoother M(diag_pa, ess_tdof_list);
```

Comment:
What do you think about the following suggestion: incorporate the call to `AssembleDiagonal` in `OperatorJacobiSmoother` so instead of
```c++
      Vector diag_pa(fespace->GetTrueVSize());
      a->AssembleDiagonal(diag_pa);
      OperatorJacobiSmoother M(diag_pa, ess_tdof_list);
```
we can just use
```
      OperatorJacobiSmoother M(a, ess_tdof_list); // call `a->AssembleDiagonal` in the constructor
```
As an added benefit, `OperatorJacobiSmoother::SetOperator` can re-compute the diagonal if the operator is changed...

Reply:
I made this change, it does make the examples a bit cleaner. However, I can't quite do what you want with `::SetOperator` because I need a `BilinearForm` (not an `Operator`) to reset the diagonal.

Reply:
The idea for `SetOperator` is that the solver class can down-cast (with `dynamic_cast`) the `Operator` to the specific class (or classes) it needs. For example, see `HypreBoomerAMG::SetOperator`.

For now, I'm fine to leave `OperatorJacobiSmoother::SetOperator` as is -- we can implement it later when needed.